### ISSUE TITLE
set() in Python gives a different ordered list everytime it is run

### ISSUE THREAD

**Issue by rigved** [hil-se/fds#39]
Hello Prof. Zhe,

In `my_AdaBoost_hint.py`, we use `self.classes_ = list(set(list(y)))`. This creates a list with a different order on each run. For example, I'll run this code with `self.n_estimators = 10`:

```python
self.classes_ = list(set(list(y)))
print('my_AdaBoost')
print(self.classes_)

for i in range(self.n_estimators):
    sample = np.random.choice(n, n, p=w)
    sampled = X.iloc[sample]
    sampled.index = range(len(sample))
    self.estimators[i].fit(sampled, labels[sample])
    print('sklearn_AdaBoost')
    print(self.estimators[i].classes_)
```

This will give the following output from two consecutive runs.

Run 1:

```bash
my_AdaBoost
['Iris-setosa', 'Iris-versicolor', 'Iris-virginica']
sklearn_AdaBoost
['Iris-setosa' 'Iris-versicolor' 'Iris-virginica']
sklearn_AdaBoost
['Iris-setosa' 'Iris-versicolor' 'Iris-virginica']
sklearn_AdaBoost
['Iris-setosa' 'Iris-versicolor' 'Iris-virginica']
sklearn_AdaBoost
['Iris-setosa' 'Iris-versicolor' 'Iris-virginica']
sklearn_AdaBoost
['Iris-setosa' 'Iris-versicolor' 'Iris-virginica']
sklearn_AdaBoost
['Iris-setosa' 'Iris-versicolor' 'Iris-virginica']
sklearn_AdaBoost
['Iris-setosa' 'Iris-versicolor' 'Iris-virginica']
sklearn_AdaBoost
['Iris-setosa' 'Iris-versicolor' 'Iris-virginica']
sklearn_AdaBoost
['Iris-setosa' 'Iris-versicolor' 'Iris-virginica']
sklearn_AdaBoost
['Iris-setosa' 'Iris-versicolor' 'Iris-virginica'
```

Run 2:

```bash
my_AdaBoost
['Iris-virginica', 'Iris-versicolor', 'Iris-setosa']
sklearn_AdaBoost
['Iris-setosa' 'Iris-versicolor' 'Iris-virginica']
sklearn_AdaBoost
['Iris-setosa' 'Iris-versicolor' 'Iris-virginica']
sklearn_AdaBoost
['Iris-setosa' 'Iris-versicolor' 'Iris-virginica']
sklearn_AdaBoost
['Iris-setosa' 'Iris-versicolor' 'Iris-virginica']
sklearn_AdaBoost
['Iris-setosa' 'Iris-versicolor' 'Iris-virginica']
sklearn_AdaBoost
['Iris-setosa' 'Iris-versicolor' 'Iris-virginica']
sklearn_AdaBoost
['Iris-setosa' 'Iris-versicolor' 'Iris-virginica']
sklearn_AdaBoost
['Iris-setosa' 'Iris-versicolor' 'Iris-virginica']
sklearn_AdaBoost
['Iris-setosa' 'Iris-versicolor' 'Iris-virginica']
sklearn_AdaBoost
['Iris-setosa' 'Iris-versicolor' 'Iris-virginica']
```

In run 1, the order in the list is the same as that of the base classifiers. However, in run 2, the order is different.

Now, if I run `print(self.estimators[i].predict_proba(X))` for any test dataset, the probabilities will have class label columns in the same order as the `sklearn` base classifier, instead of the class label columns in the order of `my_AdaBoost_hint.py`. Thus, the ensemble classifier (`my_AdaBoost_hint.py`) of run 2 will output incorrect probabilities because column 0 in the ensemble classifier is `Iris-virginica` instead of the column 0 in the base classifier (`sklearn.tree.DecisionTreeClassifier`) that refers to `Iris-setosa`.

Further, this column name ordering is assumed in the following line from the `def predict(self, X)` line of `my_AdaBoost_hint.py` file:

```python
predictions = [self.classes_[np.argmax(prob)] for prob in probs.to_numpy()]
```

One solution to this problem is to use the following code to generate `self.classes_`:

```python
self.classes_ = y.unique().tolist()
```

This code will always generate the class label list in the same order as `sklearn`'s classifiers.

What are your thoughts on this, Prof.?

Note: This problem will arise for every ensemble learning algorithm where we use a base classifier from `sklearn`.

**Comment by azhe825** [comment]
1. list(set(list(y))) should always give you the same order as long as y stays the same. Thus the class order of my_AdaBoost should always be the same. I cannot reproduce your finding.
2. However, the class order of base estimators can change (and can be different) because of the boosting process.
3. The mismatch between class orders does not affect anything since we are only using the predict function from the base estimators and the predict_proba function outputs a pd.DataFrame with class names as column headers (different from sklearn).

**Comment by rigved** [comment]
Thank you for the clarification Prof.
